src/graphemebreak.c

```diff
@@ -0,0 +1,219 @@
+ *    not be misrepresented as being the original software.
+ * 3. This notice may not be removed or altered from any source
+ *    distribution.
+ *
+ * The main reference is Unicode Standard Annex 29 (UAX #29):
+ *      <URL:http://unicode.org/reports/tr29>
+ *
+ * When this library was designed, this annex was at Revision 29, for
+ * Unicode 9.0.0:
+ *
+ * The Unicode Terms of Use are available at
+ *      <URL:http://www.unicode.org/copyright.html>
+ */
+
+/**
+ * @file    graphemebreak.c
+ *
+ * Implementation of the grapheme breaking algorithm as described in Unicode
+ * Standard Annex 29.
+ *
+ * @version 1.0, 2016
+ * @author  Andreas Roever
+ */
+
+#include "unibreakdef.h"
+#include "graphemebreak.h"
+#include "graphemebreakdata.c"
+#include "stdbool.h"
+
+#define ARRAY_LEN(x) (sizeof(x) / sizeof(x[0]))
+
+/**
+ * Initializes the wordbreak internals.  It currently does nothing, but
+ * it may in the future.
+ */
+void init_graphemebreak(void)
+{
+}
+
+/**
+ * Gets the grapheme breaking class of a character.
+ *
+ * @param ch   character to check
+ * @return     the grapheme breaking class if found; \c GBP_Other otherwise
+ */
+static enum GraphemeBreakClass get_char_gb_class(utf32_t ch)
+{
+    int min = 0;
+    int max = ARRAY_LEN(gb_prop_default) - 1;
+    int mid;
+
+    do
+    {
+        mid = (min + max) / 2;
+
+        if (ch < gb_prop_default[mid].start)
+            max = mid - 1;
+        else if (ch > gb_prop_default[mid].end)
+            min = mid + 1;
+        else
+            return gb_prop_default[mid].prop;
+    }
+    while (min <= max);
+
+    return GBP_Other;
+}
+
+/**
+ * Sets the grapheme breaking information for a generic input string.
+ *
+ * @param[in]  s             input string
+ * @param[in]  len           length of the input
+ * @param[out] brks          pointer to the output breaking data, containing
+ *                           #GRAPHEMEBREAK_BREAK or #GRAPHEMEBREAK_NOBREAK
+ * @param[in] get_next_char  function to get the next UTF-32 character
+ */
+static void set_graphemebreaks(
+        const void *s,
+        size_t len,
+        char *brks,
+        get_next_char_t get_next_char)
+{
+    size_t posNext = 0;
+    bool rule10Left = false; // is the left side of rule 10 fulfilled?
+    bool EvenRegionalIndicators = true; // is the number of preceeding GBP_RegionalIndicator characters even
+
+    utf32_t ch = get_next_char(s, len, &posNext);
+    enum GraphemeBreakClass a = get_char_gb_class(ch);
+
+    ch = get_next_char(s, len, &posNext);
+    while (ch != EOS)
+    {
+        enum GraphemeBreakClass b = get_char_gb_class(ch);
+
+        if (a == GBP_E_Base || a == GBP_E_Base_GAZ)
+          rule10Left = true;
+        else if (a != GBP_Extend)
+          rule10Left = false;
+
+        if (a == GBP_Regional_Indicator)
+          EvenRegionalIndicators = !EvenRegionalIndicators;
+        else
+          EvenRegionalIndicators = true;
+
+        if (a == GBP_CR && b == GBP_LF)
```

Comment:
I think this set of conditionals along with the few above should be moved into a "switch" statement like linebreak and wordbreak. I personally think this makes the code look clearer, but also it has the advantage that the compiler could inform you when you forget to check legal enum values in the switch(). Clang for example does that.